src/JestExt.ts

```diff
@@ -59,119 +61,95 @@ export class JestExt {
     this.testResultProvider = new TestResultProvider()
     this.debugCodeLensProvider = new DebugCodeLensProvider(this.testResultProvider, pluginSettings.enableCodeLens)
 
+    this.jestProcessManager = new JestProcessManager({
+      projectWorkspace: workspace,
+      runAllTestsFirstInWatchMode: this.pluginSettings.runAllTestsFirst,
+    })
+
     this.getSettings()
   }
 
-  public startProcess() {
-    // The Runner is an event emitter that handles taking the Jest
-    // output and converting it into different types of data that
-    // we can handle here differently.
-    if (this.jestProcess) {
-      this.jestProcess.closeProcess()
-      delete this.jestProcess
-    }
-
-    let maxRestart = 4
-    this.jestProcess = new Runner(this.workspace)
+  private handleStdErr(error: Buffer) {
+    const message = error.toString()
 
-    this.jestProcess
-      .on('debuggerProcessExit', () => {
-        this.channel.appendLine('Closed Jest')
+    if (this.shouldIgnoreOutput(message)) {
+      return
+    }
 
-        if (this.forcedClose) {
-          this.forcedClose = false
-          return
-        }
+    // The "tests are done" message comes through stdErr
+    // We want to use this as a marker that the console should
+    // be cleared, as the next input will be from a new test run.
+    if (this.clearOnNextInput) {
+      this.clearOnNextInput = false
+      this.parsingTestFile = false
+      this.testsHaveStartedRunning()
+    }
+    // thanks Qix, http://stackoverflow.com/questions/25245716/remove-all-ansi-colors-styles-from-strings
+    const noANSI = message.replace(/[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, '')
+    if (noANSI.includes('snapshot test failed')) {
+      this.detectedSnapshotErrors()
+    }
 
-        if (maxRestart-- <= 0) {
-          console.warn('jest has been restarted too many times, please check your system')
-          status.stopped('(too many restarts)')
-          return

```

Comment:
We're missing this behavior in the PR. When the Jest process in watch mode fails:
* the status bar spinner will keep spinning with "Starting watch mode" shown
* there's no annunciation on the output channel

Reply:
This is good point. The user should be notified. I will add this behaviour.

Reply:
Working on it - when the restart threshold has been reached the callback will be called so that we can update the channel and the status bar appropriately.

Reply:
Ok, so I suggest the following. For processes in watch mode I call exit callback only if they failed for a number of times (`JestProcesses.keepAliveLimit`). In the response to this, I update the channel with text *Starting Jest in Watch mode failed too many times and has been stopped. Please check your system configuration.* and I set the status bar to `status.stopped()`. This way the user will see that jest is no longer running and should start jest again. We also do not have additional callback and just use one. I will still rebase everything (if needed), and give it a spin locally today to see if nothing unexpected happens.

Reply:
Hmm...there is still one issue with this. When stopping Jest explicitly, we will get a misleading message in the channel. I suppose I will need to add an argument to the stop handler to mark that it is the result of explicit user action.

Reply:
ok - there is a better way without adding any arguments. Will fix that in a moment.

Reply:
Ok, maybe there is a better way to accomplish that without adding any extra argument to the handlers. Let me check.

Reply:
Ok, done.

Reply:
One more small thing. Just to see how it feels, on a separate branch I created a version where `JestProcess.stop` returns a promise. I tested it out when updating snapshots and it feels quite nice. I remember that @connectdotz proposed more synchronous flow. If you think it is better like this, I will probably add a test for this. Let me know. Here is the link to the commit on a separate branch: https://github.com/marcinczenko/vscode-jest/commit/5727128a8e0c3557eb40b2024d138ec6801802a8

Reply:
@seanpoulter I am running it now for two days using the version where `JestProcess.stop` returns a promise and it feels quite good. Never had any surprise. I will merge the "promise" version to the PR and I hope we will be able to merge it quickly (I would like to avoid another rebasing...). I could then proceed to issue #215 as this effectively prevents snapshots from working (I am testing snapshots by modifying the `shouldIgnoreOutput()` function).

Reply:
Sounds good. I'll wrap up my own dev efforts then focus on the review. I'm pretty short on spare time these days, so sorry in advance for it taking a while. If anyone else has the bandwidth, feel free to take the lead.